src/Chapter1.hs

```diff
@@ -187,422 +181,419 @@ ghci> :q
-
-{- |
-=⚔️= Task 8
-
-Implement a function that checks whether a given character is a vowel.
-
-🕯 HINT: use guards
-
->>> isVowel 'a'
-True
->>> isVowel 'x'
-False
--}
-isVowel c = error "isVowel: not implemented!"
-
-
-{- |
-== Local variables and functions
-
-So far, we've been playing only with simple expressions and function
-definitions. However, in some cases, expressions may become complicated, and it
-could make sense to introduce some helper variables.
-
-You can use the let-in construction in Haskell to define variables.
-Here goes an example:
-
-@
-half :: Int -> Int
-half n = let halfN = div n 2 in halfN
-@
-
-♫ NOTE: __let-in__ is also an expression! You can't just define variables; you
-  also need to return some expression that may use defined variables.
-
-The syntax for defining multiple variables requires to care about indentation
-more, but there is nothing special in it as well:
-
-@
-halfAndTwice :: Int -> (Int, Int)
-halfAndTwice n =
-    let halfN = div n 2
-        twiceN = n * 2
-    in (halfN, twiceN)
-@
-
-In addition to let-in (or sometimes even alternatively to let-in) you can use
-the __where__ construction to define local variables and functions.
-And, again, the example:
-
-@
-pythagoras :: Double -> Double -> Double
-pythagoras a b = square a + square b
-  where
-    square :: Double -> Double
-    square x = x ^ 2
-@
-
-You can define multiple functions inside __where__!
-Just remember to keep proper indentation.
--}
+closestToZero x y = if (abs x) < (abs y) then x else y
+
+-- |
+-- =⚔️= Task 7
+-- Write a function that returns the middle number among three given numbers.
+--
+-- >>> mid 3 1 2
+-- 2
+--
+-- 🕯 HINT: When checking multiple conditions, it is more convenient to use the
+--   language construction called "guards" instead of multiple nested 'if-then-else'
+--   expressions. The syntax of guards is the following:
+--
+-- @
+-- sign :: Int -> Int
+-- sign n
+--     | n < 0 = (-1)
+--     | n == 0 = 0
+--     | otherwise = 1
+-- @
+--
+-- You define different conditions in different branches, started by the '|'
+-- symbol. And the functions check them from top to bottom, returning the first
+-- value after "=" where the condition is true.
+--
+-- ♫ NOTE: The "=" sign goes after each branch, respectively.
+--
+-- ♫ NOTE: It is essential to have the same indentation before each branch "|"!
+--   Remember, that Haskell is indentation- and layout-sensitive language.
+--
+-- Casual reminder about adding top-level type signatures for all functions :)
+mid :: Int -> Int -> Int -> Int
+mid x y z
+  | x < y && x < z = if y < z then y else z -- when x is the min, decide between the other two
+  | y < x && y < z = if x < z then x else z -- when y is the min, ...
+  | z < x && z < y = if x < y then x else y -- when z is the min, ...
+  | x == y = x -- if two values are equal, one of the equal ones will be the mid
+  | y == z = y -- if two values are equal, ...
+  | z == x = z -- if two values are equal, ...
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.

Because of that, you will need to add another guard – `| otherwise = z`, to tell the compiler, that your pattern matching is exhaustive 🙂


It looks like you can move all last cases under a single otherwise and return z in here.
The implementation will be a bit shorter 🙂

Also, if you use `<=` and `>=` instead of `<` and `>`, you can have fewer guards in this implementation :slightly_smiling_face:

Reply:
I've added the otherwise, I'll try to add in the `<=` and `>=` next